lib/Transforms/LLZKInlineStructsPass.cpp

```diff
@@ -69,29 +69,63 @@ using SrcStructFieldToCloneInDest = std::map<StringRef, DestCloneOfSrcStructFiel
 using DestToSrcToClonedSrcInDest =
     DenseMap<DestFieldWithSrcStructType, SrcStructFieldToCloneInDest>;
 
-static Value getSelfValue(FuncDefOp f) {
+/// Get the "self" value for the given `FuncDefOp`, which must be either a "compute" or
+/// "constrain" function.
+static inline Value getSelfValue(FuncDefOp f) {
   if (f.nameIsCompute()) {
     return f.getSelfValueFromCompute();
   } else if (f.nameIsConstrain()) {
     return f.getSelfValueFromConstrain();
   } else {
-    llvm_unreachable("expected \"compute\" or \"constrain\" function");
+    llvm_unreachable("expected \"@compute\" or \"@constrain\" function");
   }
 }
 
+/// Get the `FieldDefOp` that defines the field referenced by the given `FieldRefOpInterface` with
+/// an assertion failure if it is not found.
 static inline FieldDefOp getDef(SymbolTableCollection &tables, FieldRefOpInterface fRef) {
   auto r = fRef.getFieldDefOp(tables);
   assert(succeeded(r));
   return r->get();
 }
 
+/// Find the `FieldWriteOp` that writes the given subcomponent struct `Value`. Produce an error
+/// (using the given callback) if there is not exactly once such `FieldWriteOp`.
+static FailureOr<FieldWriteOp>
+findOpThatStoresSubcmp(Value writtenValue, function_ref<InFlightDiagnostic()> emitError) {
+  FieldWriteOp foundWrite = nullptr;
+  for (Operation *user : writtenValue.getUsers()) {
+    if (FieldWriteOp writeOp = llvm::dyn_cast<FieldWriteOp>(user)) {
+      // Find the write op that stores the created value
+      if (writeOp.getVal() == writtenValue) {
+        if (foundWrite) {
+          // Note: There is no reason for a subcomponent to be stored to more than one field.
```

Comment:
It seems we don't have this as a semantic check, since code like this is produced by the zirgen frontend and doesn't cause issues in the up-to-date `llzk-opt.` We might want to consider either adding one (if we want this to always be enforced) or adding a transformation pass for it.